[Localization] More message ids - final batch

vmarta opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description
Ticket: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07AH000000OdGvYAK/view
Related PR: #207 

This is the final batch of message ids. After this PR is merged, I'll pass the English translation file to our translators.

Yes, this PR is huge 😅 Most of the changes are adding the message ids. But there are some interesting changes that I've pointed out in the comments below.

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [x] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [x] Changes include a UI text update in the Retail React App (which requires translation)

alexvuong committed:
add ids to actionCard and DrawerMenu

vmarta committed:
Trying some components and a page

alexvuong committed:
replace button ui type to action on messages inside action cards, add ids to some fields components

alexvuong committed:
Merge branch 'W-9467811-add-ids-to-messages' of github.com:SalesforceCommerceCloud/pwa-kit into W-9467811-add-ids-to-messages

alexvuong committed:
add instructions on message ids to readme, extract and compile messages

alexvuong committed:
Merge branch 'develop' into W-9467811-add-ids-to-messages

alexvuong committed:
compiled messages

alexvuong committed:
Merge branch 'develop' into W-9467811-add-ids-to-messages

alexvuong committed:
extracted and compiled messages

alexvuong committed:
add ids to useAddressFields component, extracted and compiled messages

vmarta committed:
Define an error message that is translatable

vmarta committed:
Message ids for Product List

vmarta committed:
Message ids for empty results

vmarta committed:
Message ids for more components

alexvuong committed:
PR feedback

alexvuong committed:
extract and translation messages

vmarta committed:
Order Summary component

vmarta committed:
Merge branch 'W-9467811-add-ids-to-messages' into more-message-ids

alexvuong committed:
PR feedback

alexvuong committed:
Merge branch 'develop' into W-9467811-add-ids-to-messages

alexvuong committed:
build translations

vmarta committed:
Merge branch 'W-9467811-add-ids-to-messages' into more-message-ids

vmarta committed:
More message ids for components

vmarta committed:
Rebuild translations

vmarta committed:
Alex feedback

vmarta committed:
Message ids for form fields

vmarta committed:
Rebuild translations

vmarta committed:
Merge branch 'develop' into more-message-ids

vmarta committed:
Rebuild translations

vmarta committed:
Message ids for account pages

vmarta committed:
Message ids for more pages

vmarta committed:
Fix syntax error

vmarta committed:
Message ids for more pages

vmarta committed:
Message ids for checkout-related

vmarta committed:
Remove this translation and hardcode it instead

Because the page is not meant to be used/seen.

vmarta committed:
Message ids for more pages

vmarta committed:
Message ids for other types of files

vmarta committed:
Message ids for the remaining components

vmarta committed:
Message ids for LocaleText component

vmarta committed:
Rebuild the translation files

vmarta commented on packages/pwa/app/components/confirmation-modal/index.test.js:
```diff
@@ -10,18 +10,7 @@ import {Box, useDisclosure} from '@chakra-ui/react'
-    alternateActionLabel: defineMessage({defaultMessage: 'No, keep item'}),
-    onPrimaryAction: noop
-}
+import {REMOVE_CART_ITEM_CONFIRMATION_DIALOG_CONFIG} from '../../pages/account/constant'
```
There's no need to define new messages from within a test file, because they were not meant to be shown in public. So I replaced it with this existing config variable.

alexvuong replied:
Since this is a test for the confirmation modal, maybe `CONFIRMATION_DIALOG_DEFAULT_CONFIG` might be more fit? 🤔

vmarta replied:
I was thinking that too. But based on the deleted messages, the test wanted the remove-cart-item version.

vmarta commented on packages/pwa/app/components/locale-text/index.jsx:
```diff
@@ -7,25 +7,22 @@
     const Wrapper = as ? chakra(as) : Text
-    const message = LOCALE_MESSAGES[shortCode] || FALLBACK_MESSAGE
+    const message = LOCALE_MESSAGES[shortCode]
 
-    if (message === FALLBACK_MESSAGE) {
+    if (!message) {
         console.error(
             `No locale message found for "${shortCode}". Please update the list accordingly.`
         )
+        return <Wrapper {...otherProps}>Unknown {shortCode}</Wrapper>
```
Now just hardcoding this fallback. I don't want it to be translated because it's meant to gracefully handle the error and for notifying the developers to fix it.

vmarta commented on packages/pwa/app/components/pagination/index.jsx:
```diff
@@ -88,11 +88,13 @@ const Pagination = (props) => {
+                            defaultMessage: 'of {numOfPages}'
+                        },
+                        {numOfPages: urls.length}
+                    )}
```
There are times when it's more readable for the translators if we combine things together like this. Otherwise, the translators would see "or" and not sure how to translate it.

vmarta commented on packages/pwa/app/components/promo-popover/index.jsx:
```diff
@@ -49,9 +49,7 @@ const PromoPopover = ({header, children, ...props}) => {
                         position="relative"
                         variant="unstyled"
-                    >
-                        <FormattedMessage defaultMessage="Applied promotions info" />
```
Deleting unnecessary children. Chakra's IconButton does not accept children.

vmarta commented on packages/pwa/app/pages/checkout/partials/shipping-address-selection.jsx:
```diff
@@ -66,14 +69,7 @@ const ShippingAddressEditForm = ({
-                                            {
-                                                defaultMessage: '{submitButtonLabel}'
-                                            },
-                                            {
-                                                submitButtonLabel
-                                            }
-                                        )}
+                                        {formatMessage(submitButtonLabel)}
```
The original message (`defaultMessage: '{someVariable}'`) would not be able to be translated because there is no hardcoded text provided. So I refactored it such that `submitButtonLabel` prop is now an actual message. You can now pass in something that makes sense to the translators: `{defaultMessage: 'Save & Continue'}`.

alexvuong replied:
Dumb question, the compiled does not work if we just have a variable in the message? 🤔

vmarta commented on packages/pwa/app/pages/login-redirect/index.jsx:
```diff
@@ -9,18 +9,11 @@
-                    defaultMessage: 'Login Redirect'
-                })}
-            </h1>
+            <h1 data-testid="login-redirect-page-heading">Login Redirect</h1>
```
We can safely hardcode this text because this LoginRedirect page won't be visible to the public.

vmarta commented on packages/pwa/app/pages/product-list/index.jsx:
```diff
@@ -164,15 +162,15 @@ const ProductList = (props) => {
-                    {defaultMessage: '{errorMessage}'},
-                    {errorMessage: API_ERROR_MESSAGE}
-                ),
+                title: formatMessage(API_ERROR_MESSAGE),
```
This is another example of when the message is just a variable name, which is something that can't be translated. So I refactored it so that `API_ERROR_MESSAGE` is now a message like this `{defaultMessage: 'Something went wrong. Try again!'}`

vmarta committed:
Update CHANGELOG.md

alexvuong commented on packages/pwa/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## v1.3.0-dev (Nov 18, 2021)
+
+-   Add message ids to all the translated text, so they provide context for the translators [#239](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/239)
```
a small ask: we should also include the other two PRs for messages Ids too. https://github.com/SalesforceCommerceCloud/pwa-kit/pull/207 and https://github.com/SalesforceCommerceCloud/pwa-kit/pull/195

alexvuong commented on packages/pwa/app/pages/checkout/partials/shipping-address-selection.jsx:
```diff
@@ -91,13 +87,20 @@ ShippingAddressEditForm.propTypes = {
     form: PropTypes.object,
-    submitButtonLabel: PropTypes.string
+    submitButtonLabel: PropTypes.shape({
+        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
```
```suggestion
        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
```

alexvuong commented on packages/pwa/app/components/forms/form-action-buttons.jsx:
```diff
@@ -42,8 +46,12 @@ const FormActionButtons = ({
-    saveButtonLabel: PropTypes.string,
-    cancelButtonLabel: PropTypes.string,
+    saveButtonLabel: PropTypes.shape({
+        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
```
```suggestion
        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
```

alexvuong commented on packages/pwa/app/components/forms/form-action-buttons.jsx:
```diff
@@ -42,8 +46,12 @@ const FormActionButtons = ({
+        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
+    }),
+    cancelButtonLabel: PropTypes.shape({
+        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
```
Same as above

alexvuong commented on packages/pwa/app/components/forms/useCreditCardFields.jsx:
```diff
@@ -8,16 +8,31 @@ import cardValidator from 'card-validator'
-    cardName: {defaultMessage: 'Name on Card'},
-    expiryDate: {defaultMessage: 'Expiry Date'},
-    securityCode: {defaultMessage: 'Security Code'}
+    required: {defaultMessage: 'Required', id: 'use_credit_card_fields.error.required'},
```
I thought I've added the ids for this component in my previous PR. 🤔 . Why are we seeing it here?
[Here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/packages/pwa/app/components/forms/useCreditCardFields.jsx#L11) is what it looks like on `develop`

alexvuong commented on packages/pwa/app/components/item-variant/item-price.jsx:
```diff
@@ -24,6 +24,7 @@ const PricePerItem = ({currency, basket, basePrice}) => {
             />
             <FormattedMessage
                 defaultMessage="ea"
+                id="price_per_item.label.each"
```
Nit: the location `price_per_item` is not consistent with the component name. I find `item-price` makes sense. Is there a reason why `price_per_item` is used instead?

vmarta replied:
Because this one lives in a separate component called `PricePerItem`.

alexvuong replied:
oh I see, sorry i did not see the component there. It makes sense to me now 👍

vmarta committed:
Merge branch 'develop' into more-message-ids-2

vmarta committed:
Rebuild translations

alexvuong commented on packages/pwa/app/hooks/use-auth-modal.js:
```diff
@@ -65,7 +65,8 @@ export const AuthModal = ({
                       defaultMessage:
-                          "Something's not right with your email or password. Try again."
+                          "Something's not right with your email or password. Try again.",
+                      id: 'auth_modal.error.incorrect_email_password'
```
Nit: 
```suggestion
                      id: 'auth_modal.error.incorrect_email_or_password'
```

alexvuong commented on packages/pwa/app/pages/account/constant.js:
```diff
@@ -52,18 +52,31 @@ export const navLinks = [
-    confirmationMessage: {defaultMessage: 'Are you sure you want to continue ?'},
-    primaryActionLabel: {defaultMessage: 'Yes'},
-    alternateActionLabel: {defaultMessage: 'No'}
+    dialogTitle: {defaultMessage: 'Confirm Action', id: 'global.account.title.confirm_action'},
```
```suggestion
    dialogTitle: {defaultMessage: 'Confirm Action', id: 'confirmation_dialog.title.confirm_action'},
```
The constant name sounds to me it belongs to the `confirmation_dialog` component instead of a global/account pages

alexvuong commented on packages/pwa/app/pages/account/constant.js:
```diff
@@ -52,18 +52,31 @@ export const navLinks = [
-    dialogTitle: defineMessage({defaultMessage: 'Confirm Remove Item'}),
+    dialogTitle: defineMessage({
+        defaultMessage: 'Confirm Remove Item',
+        id: 'global.account.title.confirm_remove'
```
```suggestion
        id: 'cart_item.confirmation_dialog.title.confirm_remove'
```
Same with above, the name already indicates which component it belongs to, it makes sense to use it for the ids.

alexvuong commented on packages/pwa/app/pages/account/index.jsx:
```diff
@@ -74,7 +74,8 @@ const Account = () => {
                         {formatMessage({
-                            defaultMessage: 'Log out'
+                            defaultMessage: 'Log out',
+                            id: 'account.logout_button.button.log_out'
```
What do you think of this id? 
```suggestion
                            id: 'account_page.button.log_out'
```

vmarta replied:
I think I prefer to have a more specific hierarchy in this case: Account page → LogoutButton component

alexvuong replied:
oh I see, sorry I did not see there is a LogOut button there

alexvuong commented on packages/pwa/app/pages/account/wishlist/partials/wishlist-secondary-button-group.jsx:
```diff
@@ -18,12 +18,22 @@ import {noop} from '../../../../utils/utils'
-    dialogTitle: defineMessage({defaultMessage: 'Confirm Remove Item'}),
+    dialogTitle: defineMessage({
+        defaultMessage: 'Confirm Remove Item',
+        id: 'global.wishlist.title.confirm_remove'
```
```suggestion
        id: 'wishlist_item.confirmation_modal.title.confirm_remove'
```

alexvuong commented on packages/pwa/app/pages/account/wishlist/partials/wishlist-secondary-button-group.jsx:
```diff
@@ -18,12 +18,22 @@ import {noop} from '../../../../utils/utils'
 import {API_ERROR_MESSAGE} from '../../../../constants'
 
 export const REMOVE_WISHLIST_ITEM_CONFIRMATION_DIALOG_CONFIG = {
```
Thought: Should we move this to `constant.js` as we already have similar type of constant in there? (e.g REMOVE_CART_ITEM_CONFIRMATION_DIAGLOG_CONFIG)

vmarta replied:
I think I'll just leave it there. Looks like it's meant to be used in the same file. So I'm not sure why it was being exported. But yeah, I'll just leave it as it is :)

alexvuong commented on packages/pwa/app/pages/account/wishlist/partials/wishlist-secondary-button-group.jsx:
```diff
@@ -68,7 +81,10 @@ const WishlistSecondaryButtonGroup = ({productListItemId, onClick = noop}) => {
-                    <FormattedMessage defaultMessage="Remove" />
+                    <FormattedMessage
+                        defaultMessage="Remove"
+                        id="wishlist_secondary_button_group.action.remove"
```
Should we consider this as a button? 🤔  
```suggestion
                        id="wishlist_secondary_button_group.button.remove"
```

vmarta replied:
Since this is inside `<Button variant="link">` that has an `onClick` prop, I'd consider this as an 'action'.

alexvuong commented on packages/pwa/app/pages/cart/index.jsx:
```diff
@@ -67,7 +67,8 @@ const Cart = () => {
                         defaultMessage:
-                            '{quantity} {quantity, plural, one {item} other {items}} added to wishlist'
+                            '{quantity} {quantity, plural, one {item} other {items}} added to wishlist',
+                        id: 'cart.info.added_to_wishlist'
```
Since we do have some cart components, it is clearer to say `cart_page` for the ids here
```suggestion
                        id: 'cart_page.info.added_to_wishlist'
```

vmarta replied:
🤔 Since we don't have an actual `<Cart>` component, I think this is ok as it is.

alexvuong commented on packages/pwa/app/pages/cart/partials/cart-secondary-button-group.jsx:
```diff
@@ -23,12 +23,22 @@ import {noop} from '../../../utils/utils'
 import useCustomer from '../../../commerce-api/hooks/useCustomer'
 
 export const REMOVE_CART_ITEM_CONFIRMATION_DIALOG_CONFIG = {
```
Just check the code, we have two places that declared `REMOVE_CART_ITEM_CONFIRMATION_DIALOG_CONFIG` (one here, and another at `pages/account/constant.js`)? Can we remove one? 🤔

vmarta replied:
Thanks, Alex. I've removed one of them.

alexvuong commented on packages/pwa/app/pages/cart/partials/cart-secondary-button-group.jsx:
```diff
@@ -23,12 +23,22 @@ import {noop} from '../../../utils/utils'
-    dialogTitle: defineMessage({defaultMessage: 'Remove Item?'}),
+    dialogTitle: defineMessage({
+        defaultMessage: 'Remove Item?',
+        id: 'global.cart.title.remove_item'
```
Left a comments for the ids [above](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/239/files#r758763369)

vmarta replied:
Yeah, let's replace these 'global' ids because they're not as global as I initially thought 😅

alexvuong commented on packages/pwa/app/pages/login/index.jsx:
```diff
@@ -31,7 +31,8 @@ const Login = () => {
                 ? formatMessage({
-                      defaultMessage: 'Incorrect username or password, please try again.'
+                      defaultMessage: 'Incorrect username or password, please try again.',
+                      id: 'login.error.incorrect_username_password'
```
```suggestion
                      id: 'login_page.error.incorrect_username_password'
```

vmarta replied:
Is adding '_page' here because we want to differentiate it from the same-named Login component?

alexvuong replied:
yes, since we have a login as a component and login as a page. It would be better if we use `login_page` for this one

alexvuong commented on packages/pwa/app/pages/checkout/confirmation.jsx:
```diff
@@ -74,10 +74,16 @@ const CheckoutConfirmation = () => {
-                        <FormattedMessage defaultMessage="Log in here" />
+                        <FormattedMessage
+                            defaultMessage="Log in here"
+                            id="checkout_confirmation.link_login"
```
```suggestion
                            id="checkout_confirmation.link.login"
```

vmarta replied:
Good catch, thanks.

alexvuong commented on packages/pwa/app/pages/checkout/partials/contact-info.jsx:
```diff
@@ -63,7 +63,8 @@ const ContactInfo = () => {
                     formatMessage({
-                        defaultMessage: 'Incorrect username or password, please try again.'
+                        defaultMessage: 'Incorrect username or password, please try again.',
+                        id: 'contact_info.error.incorrect_username_password'
```
Nit:
```suggestion
                        id: 'contact_info.error.incorrect_username_or_password'
```

alexvuong commented on packages/pwa/app/pages/checkout/partials/payment-selection.jsx:
```diff
@@ -153,7 +162,10 @@ const PaymentSelection = ({form, hideSubmitButton, onSubmit = () => null}) => {
-                                                                <FormattedMessage defaultMessage="Save & Continue" />
+                                                                <FormattedMessage
+                                                                    defaultMessage="Save & Continue"
+                                                                    id="payment_selection.button.save_continue"
```
```suggestion
                                                                    id="payment_selection.button.save_and_continue"
```

alexvuong commented on packages/pwa/app/pages/checkout/partials/shipping-address-selection.jsx:
```diff
@@ -18,6 +18,11 @@ import AddressDisplay from '../../../components/address-display'
 
+const saveButtonMessage = defineMessage({
+    defaultMessage: 'Save & Continue to Shipping Method',
+    id: 'shipping_address_edit_form.button.save_continue'
```
```suggestion
    id: 'shipping_address_edit_form.button.save_and_continue'
```

alexvuong commented on packages/pwa/app/pages/checkout/partials/shipping-address-selection.jsx:
```diff
@@ -349,7 +353,9 @@ ShippingAddressSelection.propTypes = {
     /** Override the submit button label */
-    submitButtonLabel: PropTypes.string,
+    submitButtonLabel: PropTypes.shape({
+        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.array])
```
In which case it would be a string? 🤔 , afaik, `defaultMessage` has always been an object. 
```suggestion
        defaultMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
```

vmarta replied:
`defaultMessage` is inside an object. `defaultMessage` itself is written as a string. 
```js
{id: '...', defaultMessage: 'hello world'}
```
But our localization setup dynamically import a _compiled_ version of the translations. This results in `defaultMessage` in array form. Maybe I'll add a comment in the prop types to make it clearer.

vmarta committed:
Update packages/pwa/app/hooks/use-auth-modal.js

Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vmarta committed:
Update packages/pwa/app/pages/checkout/confirmation.jsx

Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vmarta committed:
Update packages/pwa/app/pages/checkout/partials/shipping-address-selection.jsx

Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vmarta committed:
Update packages/pwa/app/pages/checkout/partials/payment-selection.jsx

Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vmarta committed:
Update packages/pwa/app/pages/checkout/partials/contact-info.jsx

Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vmarta committed:
From Alex's feedback

vmarta committed:
More of Alex's feedback

- removed duplicate constant
- rename the global ids to something more specific (because they're not used globally actually))
- improved the message proptype

vmarta committed:
Remove unused references

vmarta committed:
Rebuild translations

vmarta committed:
Differentiate between login page and component

vmarta committed:
Rebuild pseudo-locale file

alexvuong commented on packages/pwa/app/pages/login/index.jsx:
```diff
@@ -31,7 +31,8 @@ const Login = () => {
                 ? formatMessage({
-                      defaultMessage: 'Incorrect username or password, please try again.'
+                      defaultMessage: 'Incorrect username or password, please try again.',
+                      id: 'login_page.error.incorrect_username_password'
```
```suggestion
                      id: 'login_page.error.incorrect_username_or_password'
```

alexvuong reviewed (approved):
LGTM, with just a small comment. It is good to merge after it is changed

vmarta committed:
Update packages/pwa/app/pages/login/index.jsx

Co-authored-by: Alex Vuong <52219283+alexvuong@users.noreply.github.com>

vmarta committed:
Rebuild translation files
